Feature exception strings browser subprocess

illfang opened the pull request:
Added a string util function for generating detailed exception strings and facilitate it in the javascript method handler of the browser subprocess

amaitland commented:
Please provide a description.

amaitland commented:
No need for both `PR's`, let's just go with this one as it already has both commits.

amaitland commented:
Can you provide an example to demo your changes?

amaitland commented on CefSharp.Core/Internals/StringUtils.cpp:
```diff
@@ -78,5 +78,20 @@ namespace CefSharp
         }
+
+        [DebuggerStepThrough]
+        String^ StringUtils::CreateExceptionString(Exception^ ex, int limit)
```
No need for the code duplication, can just go with the `C#` version in a common place.

illfang commented:
Sorry, I did not notice that this branch contained both commits. I've refactored the source code.
I've added a diff for an example of the log messages.
http://pastebin.com/Jaezue4B

The created log is:

Result of calling echoInt with double
```
Line: 186, Source: custom://cefsharp/BindingTest.html, Message: Uncaught Error: Could not execute method: echoInt(3.14159265358979) - Missing Parameters: 0:
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 126
caused by:
Object of type 'System.Double' cannot be converted to type 'System.Int32'.:
   at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast)
   at System.RuntimeType.CheckValue(Object value, Binder binder, CultureInfo culture, BindingFlags invokeAttr)
   at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfoA first chance exception of type 'System.InvalidOperationException' occurred in CefSharp.dll
 culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
```

Result of calling echoInt without parameter
```
Line: 194, Source: custom://cefsharp/BindingTest.html, Message: Uncaught Error: Could not execute method: echoInt(System.Reflection.Missing) - Missing Parameters: 1:
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 126
caused by:
Missing parameter does not have a default value.
Parameter name: parameters:
   at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
```

As you can see the newly thrown exception includes a log of the called method including the parameters and the count of missing parameters. Furthermore, the reason for the exception is also added to the exception string by recursivly traversing the inner exceptions which helps in debugging.

amaitland commented:
> Furthermore, the reason for the exception is also added to the exception string by recursivly traversing the inner exceptions which helps in debugging.

Doesn't `.Net` have this built in already? I'd have to look it up, though there was a method.

amaitland commented on CefSharp/Internals/ExceptionExtension.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <summary>
+        /// The default log count for inner exceptions.
+        /// </summary>
+        public const int INNER_EXCEPTION_LOG_COUNT = 5;
```
`.Net` style naming please, e.g.` InnerExceptionLogCount`

amaitland commented on CefSharp/Internals/ExceptionExtension.cs:
```diff
@@ -0,0 +1,40 @@
+                limit = limit - 1;
+                return ex.Message + ":\n" + ex.StackTrace + "\ncaused by:\n" + innerException.CreateDetailedExceptionString(limit);
+            }
+            else
```
Remove redundant else.

amaitland commented on CefSharp/Internals/JavascriptObjectRepository.cs:
```diff
@@ -118,7 +118,13 @@ public bool TryCallMethod(long objectId, string name, object[] parameters, out o
+                {
+                    result = method.Function(obj.Value, parameters);
+                }
+                catch (Exception e) {
```
Bracket on new line

amaitland commented on CefSharp/Internals/ExceptionExtension.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <param name="limit">The optional limit for logging recrusive inner exceptions.
+        /// Default value is: INNER_EXCEPTION_LOG_COUNT</param>
+        [DebuggerStepThrough]
+        public static String CreateDetailedExceptionString(this Exception ex, int limit = INNER_EXCEPTION_LOG_COUNT)
```
Probably better as `ToDetailedExceptionString`

amaitland replied:
Actually scrap the exception part, we all know it's an exception, so just `ToDetailedString`

amaitland commented on CefSharp.BrowserSubprocess.Core/JavascriptMethodHandler.cpp:
```diff
@@ -38,7 +38,7 @@ namespace CefSharp
         catch (Exception^ ex)
         {
-            exception = StringUtils::ToNative(ex->Message);
+            exception = StringUtils::ToNative(ExceptionExtension::CreateDetailedExceptionString(ex, ExceptionExtension::INNER_EXCEPTION_LOG_COUNT));
```
Does `VC++` not play nice with the default param?

illfang replied:
No it does not compile when removing the constant.
When adding [System::Runtime::InteropServices::DefaultParameterValue(INNER_EXCEPTION_LOG_COUNT)] to the parameter of CreateDetailedExceptionString it will compile but then the C# version does not work anymore as it is not possible to define both the annotation and the = default parameter.
Thought about method overloading but that seemed a bit excessive as it is not resolved at compile time.

amaitland replied:
Leave as is then.

amaitland commented on CefSharp/Internals/ExceptionExtension.cs:
```diff
@@ -0,0 +1,40 @@
+﻿// Copyright © 2010-2015 The CefSharp Project. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+using System;
```
Add missing line here

illfang commented:
> Doesn't .Net have this built in already? I'd have to look it up, though there was a method.
Unfortunately, I did not find such a feature in .Net.

amaitland commented:
> I have committed a cleaned up version taking your comments into account.

Much cleaner, thanks :+1:

amaitland commented:
> Unfortunately, I did not find such a feature in .Net.

Doesn't just calling `ToString()` on the exception provide basically everything you've got there. It should recursively build up the inner exceptions. See reference source

http://referencesource.microsoft.com/#mscorlib/system/exception.cs,438

illfang commented:
I found it hard to see which stack trace belongs to which excepation. However, I can understand if you want to rely on the .Net ToString method.

ToString():
```
Line: 158, Source: custom://cefsharp/BindingTest.html, Message: Uncaught Error: System.InvalidOperationException: Could not execute method: repeat(hi , 5) - Missing Parameters: 0 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.InvalidOperationException: Cannot repeat ---> System.InvalidProgramException: BAD ---> System.DivideByZeroException: Attempted to divide by zero.
   at CefSharp.Example.BoundObject.Repeat(String str, Int32 n) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\BoundObject.cs:line 78
   --- End of inner exception stack trace ---
   at CefSharp.Example.BoundObject.Repeat(String str, Int32 n) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\BoundObject.cs:line 82
   --- End of inner exception stack trace ---
   at CefSharp.Example.BoundObject.Repeat(String str, Int32 n) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\BoundObject.cs:line 85
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
   --- End of inner exception stack trace ---
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
```

ToDetailedString:
```
Line: 158, Source: custom://cefsharp/BindingTest.html, Message: Uncaught Error: Could not execute method: repeat(hi , 5) - Missing Parameters: 0:
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Exception has been thrown by the target of an invocation.:
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, OA first chance exception of type 'System.InvalidOperationException' occurred in CefSharp.dll
bject[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
caused by:
Cannot repeat:
   at CefSharp.Example.BoundObject.Repeat(String str, Int32 n) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\BoundObject.cs:line 85
caused by:
BAD:
   at CefSharp.Example.BoundObject.Repeat(String str, Int32 n) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\BoundObject.cs:line 82
caused by:
Attempted to divide by zero.:
   at CefSharp.Example.BoundObject.Repeat(String str, Int32 n) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\BoundObject.cs:line 78
```

(diff for this log http://pastebin.com/CqkFEsea )

amaitland commented:
> I found it hard to see which stack trace belongs to which excepation. However, I can understand if you want to rely on the .Net ToString method.

I'm somewhat wary of deviating from the norm here. I'm not against it, just need to make sure we're making an informed choice. Based on the output you've provided above, I'm not sure the sample you've provided is what I would typically expect. For one the ordering is different, this is likely unexpected and confusing for people expecting the standard output.

amaitland commented:
> (diff for this log http://pastebin.com/CqkFEsea )

Also rather than modifying an existing method to demo your code, can you please add another method that includes your additional `try....catch` code, just not called by `BindingTest.html`, just to make it easy to reproduce.

illfang commented:
Output of this version using the ExceptionTest of #1254 
```
Exception string for nested exceptions:
Could not execute method: triggerNestedExceptions() - Missing Parameters: 0:
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Exception has been thrown by the target of an invocation.:
at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) 
at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) 
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) 
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
caused by:
Nested Exception Canceled:
at CefSharp.Example.ExceptionTestBoundObject.TriggerNestedExceptions() in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 37
caused by:
Nested Exception Invalid:
at CefSharp.Example.ExceptionTestBoundObject.TriggerNestedExceptions() in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 33
caused by:
Attempted to divide by zero.:
at CefSharp.Example.ExceptionTestBoundObject.DivisionByZero(Int32 zero) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 19 
at CefSharp.Example.ExceptionTestBoundObject.TriggerNestedExceptions() in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 29
Stack:
Error: Could not execute method: triggerNestedExceptions() - Missing Parameters: 0:
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Exception has been thrown by the target of an invocation.:
at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) 
at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) 
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) 
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
caused by:
Nested Exception Canceled:
at CefSharp.Example.ExceptionTestBoundObject.TriggerNestedExceptions() in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 37
caused by:
Nested Exception Invalid:
at CefSharp.Example.ExceptionTestBoundObject.TriggerNestedExceptions() in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 33
caused by:
Attempted to divide by zero.:
at CefSharp.Example.ExceptionTestBoundObject.DivisionByZero(Int32 zero) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 19 
at CefSharp.Example.ExceptionTestBoundObject.TriggerNestedExceptions() in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp.Example\ExceptionTestBoundObject.cs:line 29
at Error (native)
at custom://cefsharp/ExceptionTest.html:16:47

Exception string for wrong parameter type:
Could not execute method: triggerParameterException(3.14159265358979) - Missing Parameters: 0:
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Object of type 'System.Double' cannot be converted to type 'System.Int32'.:
at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast) 
at System.RuntimeType.CheckValue(Object value, Binder binder, CultureInfo culture, BindingFlags invokeAttr) 
at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig) 
at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) 
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
Stack:
Error: Could not execute method: triggerParameterException(3.14159265358979) - Missing Parameters: 0:
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Object of type 'System.Double' cannot be converted to type 'System.Int32'.:
at System.RuntimeType.TryChangeType(Object value, Binder binder, CultureInfo culture, Boolean needsSpecialCast) 
at System.RuntimeType.CheckValue(Object value, Binder binder, CultureInfo culture, BindingFlags invokeAttr) 
at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig) 
at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) 
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
at Error (native)
at custom://cefsharp/ExceptionTest.html:27:47

Exception string for missing parameter:
Could not execute method: triggerParameterException(System.Reflection.Missing) - Missing Parameters: 1:
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Missing parameter does not have a default value. 
Parameter name: parameters:
at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig) 
at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) 
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
Stack:
Error: Could not execute method: triggerParameterException(System.Reflection.Missing) - Missing Parameters: 1:
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 127
caused by:
Missing parameter does not have a default value. 
Parameter name: parameters:
at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig) 
at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) 
at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) 
at CefSharp.Internals.JavascriptObjectRepository.TryCallMethod(Int64 objectId, String name, Object[] parameters, Object& result, String& exception) in c:\Users\kevin\Desktop\MyCefSharp\CefSharp\CefSharp\Internals\JavascriptObjectRepository.cs:line 123
at Error (native)
at custom://cefsharp/ExceptionTest.html:38:47
```

illfang commented:
I have created two version for this feature, the one discussed here using custom formatting and one using the .Net standard. The PR for the second version is #1255 
I can understand your concerns regarding the custom formatting. Please decide if you want to integrate one of the versions.

Furthermore, to reproduce the log an exception log test is available with #1254

amaitland commented:
> I have created two version for this feature, the one discussed here using custom formatting and one using the .Net standard. The PR for the second version is #1255

I've merged #1255, if you feel there is merit in providing a custom exception mess then I'm happy to discuss this further.

> Furthermore, to reproduce the log an exception log test is available with #1254

Thanks

illfang commented:
> if you feel there is merit in providing a custom exception mess then I'm happy to discuss this further.

Personally, I think the log order of this request is clearer because it is not necessary to count which trace belongs to which exception.
However, I believe this is determined by personal preference and will therefore close this request.
